engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailAggregator.java

```diff
@@ -142,31 +177,107 @@ public Map<User, Map<String, Object>> getAggregated(UUID appId, EmailAggregation
             totalAggregatedElements += aggregations.size();
         } while (maxPageSize == aggregations.size());
         Log.infof("%d elements were aggregated for key %s", totalAggregatedElements, aggregationKey);
+    }
 
-        return aggregated.entrySet().stream()
-                .filter(entry -> !entry.getValue().isEmpty())
-                .collect(toMap(
-                        Entry::getKey,
-                        entry -> entry.getValue().getContext()
-                ));
+    private void aggregationBasedOnEvent(EventAggregationCriteria eventAggregationCriteria, LocalDateTime start, LocalDateTime end, Map<String, Set<String>> subscribersByEventType, Map<String, Set<String>> unsubscribersByEventType, Map<User, AbstractEmailPayloadAggregator> aggregated) {
+        int offset = 0;
+        int totalAggregatedElements = 0;
+
+        List<Event> aggregations;
+        do {
+            // First, we retrieve paginated aggregations that match the given key.
+            aggregations = emailAggregationRepository.getEmailAggregationBasedOnEvent(eventAggregationCriteria, start, end, offset, maxPageSize);
+            offset += maxPageSize;
+
+            // For each aggregation...
+            for (Event aggregation : aggregations) {
+                // We need its event type to determine the target endpoints.
+                EventType eventType = aggregation.getEventType();
```

Comment:
Doesn't that line currently trigger one additional SQL query for each event? If so, we should probably use `JOIN FETCH` in `EmailAggregationRepository`.

Reply:
Sure, done